### PR TITLE
unit testing parser and git modules 

### PR DESCRIPTION
**Pull request by malialiu**
Unit tests for git_checkout and parse_gitlog

### CODE REVIEW

File: vignettes/download_jira_data.Rmd [hunk 1]
```diff
@@ -17,6 +17,12 @@ seed <- 1
 set.seed(seed)
 ```
 
+Some packages might not be downloadable by install.packages(“packages_name”), we can try to install the package directly from the github
```
Comment (carlosparadis) [inline vignettes/download_jira_data.Rmd]: This seems helpful info in general, rather than specific to this notebook regarding installing R package. Could you add this to https://github.com/sailuh/kaiaulu/wiki/Dependencies-Troubleshooting instead?

File: vignettes/download_jira_data.Rmd [hunk 2]
```diff
@@ -35,7 +41,7 @@ As usual, the first step is to load the project configuration file.
 # Project Configuration File
 
 ```{r}
-conf <- yaml::read_yaml("../conf/geronimo.yml")
+conf <- yaml::read_yaml("../conf/thrift.yml")
```
Comment (carlosparadis) [inline vignettes/download_jira_data.Rmd]: Please keep the original geronimo config, since the notebook discusses the properties of the data associated to it.

File: vignettes/social_smell_showcase.Rmd [hunk 3]
```diff
@@ -32,6 +32,9 @@ The bottom line is, the required effort to obtain the mailing list data will var
 # Libraries
 
 Please ensure the following R packages are installed on your computer. 
+On Ubuntu the libfontconfig1-dev package and libharfbuzz-dev package are required. It not, here are some command recommended as the following:
```
Comment (carlosparadis) [inline vignettes/social_smell_showcase.Rmd]: This is very helpful for the wiki on Troubleshooting Dependencies. Could you add to it instead of the notebook?

File: vignettes/social_smell_showcase.Rmd [hunk 4]
```diff
@@ -56,7 +59,7 @@ We also provide the path for `tools.yml`. Kaiaulu does not implement all availab
 
 ```{r}
 tools_path <- "../tools.yml"
-conf_path <- "../conf/helix.yml"
+conf_path <- "../conf/thrift.yml"
```
Comment (carlosparadis) [inline vignettes/social_smell_showcase.Rmd]: Please keep the original helix file here too.

File: vignettes/social_smell_showcase.Rmd [hunk 5]
```diff
@@ -94,7 +97,16 @@ As stated in the introduction, we need both git log and at least one communicati
 
 ## Parse Gitlog
 
+Recommended perceval version: 0.12.24
```
Comment (carlosparadis) [inline vignettes/social_smell_showcase.Rmd]: This is specific to Kaiaulu, rather than just this notebook, so let's leave this one out since one PR already adds it to the README (or at least one issue I have that noted down).

File: vignettes/social_smell_showcase.Rmd [hunk 6]
```diff
@@ -94,7 +97,16 @@ As stated in the introduction, we need both git log and at least one communicati
...
 To get started, we use the `parse_gitlog` function to extract a table from the git log. You can inspect the `project_git` variable to inspect what information is available from the git log. 
+Reminder: the parse_gitlog function has a path.expand function, Expand paths (e.g. "~/Desktop" => "/Users/someuser/Desktop")
```
Comment (carlosparadis) [inline vignettes/social_smell_showcase.Rmd]: This is good docs, but this should be added directly to the function parse_gitlog instead of the notebook, above the function definition on R/parse.R. Could you move this there instead?

File: vignettes/social_smell_showcase.Rmd [hunk 7]
```diff
@@ -97,16 +94,7 @@ As stated in the introduction, we need both git log and at least one communicati
 
 ## Parse Gitlog
 
-Recommended perceval version: 0.12.24
 To get started, we use the `parse_gitlog` function to extract a table from the git log. You can inspect the `project_git` variable to inspect what information is available from the git log. 
-Reminder: the parse_gitlog function has a path.expand function, Expand paths (e.g. "~/Desktop" => "/Users/someuser/Desktop")
-Example paths on Ubuntu:
```
Comment (carlosparadis) [inline vignettes/social_smell_showcase.Rmd]: I see you removed this, but I'd still encourage a commit to update the docs on parse_gitlog since this may help others. Likewise, with moving some of the other docs to the wiki on the page I added.

File: tests/unit_test.R [hunk 8]
```diff
@@ -0,0 +1,38 @@
+library(kaiaulu)
+library(testthat)
+
+test_that("calling parse_gitlog with correct perceval path and an exist local path of github project", {
+  git_repo_path <- "~/Documents/GitHub/thrift/.git"
```
Comment (carlosparadis) [inline tests/unit_test.R:5]: This does not follow the agreed convention on Kaiaulu where rawdata is. Maybe check with @leilani-reich on this.

File: tests/unit_test.R [hunk 9]
```diff
@@ -0,0 +1,38 @@
+library(kaiaulu)
+library(testthat)
+
+test_that("calling parse_gitlog with correct perceval path and an exist local path of github project", {
```
Comment (carlosparadis) [inline tests/unit_test.R:4]: This is a very good start for the description. It has a good amount of detail, but it is missing the behavior on the sentence. "test that calling parse_ghitlog with correct perceval path **returns a data table**). 

(Note I specified data.table, since that's the object parse_gitlog returns, using the data.table package, which is different than R's native dataframe).

File: tests/unit_test.R [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+  git_repo_path <- "~/Documents/GitHub/thrift/.git"
+  perceval_path <- "/home/mliu0141/.local/bin/perceval"
```
Comment (carlosparadis) [inline tests/unit_test.R:6]: This also does not follow the agreed convention in Kaiaulu.

File: tests/unit_test.R [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+  result <- parse_gitlog(perceval_path, git_repo_path)
```
Comment (carlosparadis) [inline tests/unit_test.R:8]: One problem you will run into here is on the assumption the user has the data, even if the paths follow the convention. Similar issue @leilani-reich functions will run into for these. More on this in a follow-up.

File: tests/unit_test.R (hunk 8 above)
Comment (carlosparadis) [reply tests/unit_test.R:5]: Please apply the fixes to the subsequent tests too.

File: tests/unit_test.R (hunk 9 above)
Comment (carlosparadis) [reply tests/unit_test.R:4]: Please apply the fix to the description to the subsequent tests too.

**Review by carlosparadis — CHANGES_REQUESTED** [review]
The tests are good in size, but there are a few things that needs to be fixed across them. 

For starters, please post a bullet list of what you want to test in plain English on this pull request. Use this as an exercise for my comment on the description of the tests.

**Comment by carlosparadis** [comment]
Please remember to include these pull requests on the board (right menu), under review, when you update here, or I may run the risk of not noticing there are more to be reviewed. @leilani-reich can help you with this if you are not clear how. I just added it to the backlog for you.

File: vignettes/social_smell_showcase.Rmd (hunk 6 above)
Comment (malialiu) [reply vignettes/social_smell_showcase.Rmd]: This is already in there, I added to the showcase as reminder because I think people might not know that parser_gitlog function will expand the path when they running the showcase.

File: vignettes/social_smell_showcase.Rmd (hunk 7 above)
Comment (malialiu) [reply vignettes/social_smell_showcase.Rmd]: I was planning on putting something like this
```{r}
Reminder: the parse_gitlog function has a path.expand function, Expand paths (e.g. "~/Desktop" => "/Users/someuser/Desktop")
Example paths on Ubuntu:
perceval_path: /home/myComputer/.local/bin/perceval
git_repo_path: ~/Documents/GitHub/thrift/.git

Example paths on Windows: (there seems to be some issues to run perceval in Windows)
perceval_path: C:/Users/myComputer/AppData/Local/Packages/PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0/LocalCache/local-packages/Python310/Scripts/perceval
git_repo_path: C:/Users/myComputer/Documents/GitHub/thrift/.git
```
But I think perceval will not work in Windows so I removed it and the parse_gitlog already has this in doc
```{r}
# Expand paths (e.g. "~/Desktop" => "/Users/someuser/Desktop")
```

**Comment by carlosparadis** [comment]
Hi @malialiu 

I added a few more comments, but you are almost there on the unit tests. When you make the next commit, please git squash them all into one commit.

Also, you want to separate your tests into the appropriate (new) files:

`git_checkout` tests should exist in a `test-git.R` file, whereas the `parser_gitlog` tests should exist in a `test-parser.R` file. The file names should match the file names of where the function belongs.

**Comment by carlosparadis** [comment]
Hi,

You should be careful about merging commits after I make code revisions and before you implement them. Note my review now lies in an "outdated" category. If it is clear to you what is the change request, that is fine.

File: tests/testthat/test-git.R [hunk 12]
```diff
@@ -0,0 +1,23 @@
+test_that("test that calling git_checkout with correct branch name and an exist local path of github project returns list of string", {
```
Comment (carlosparadis) [inline tests/testthat/test-git.R:1]: Hi Malia,

I am not sure if I am missing something here, but some of the past revision requests are still pending on your tests. For example, I still see the tests are "test _that (test that)". I recall mentioning this before.

**Review by carlosparadis — CHANGES_REQUESTED** [review]
See above. I think something broke on the review requests during one of the force pushes.

**Comment by carlosparadis** [comment]
Maybe all that is missing is the test_that("test that"). Just remove the extra "test that" from the string. :)

**Comment by leilani-reich** [comment]
Hi @carlosparadis, I have not assigned Nico to Malia's unit tests because she says she thinks all the revisions are done. Or should I anyway? Thanks.

**Comment by carlosparadis** [comment]
I am not sure I understand what "all revisions are done means". Did Malia already do the 3 Asks we discussed? I am planning to read them later today to confirm, but if not, then she still has work to do on this Pull Request, no?

**Comment by leilani-reich** [comment]
This is the unit testing for milestone 1. You asked me to assign the milestone 1 pending pull requests to Nico, right?

**Comment by carlosparadis** [comment]
Are all yours, Malia's, and Nico's unit tests included in this Pull Request that is listed in https://github.com/sailuh/kaiaulu/issues/154#issue-1597495593 (and or any other unit tests of milestone 1)?

**Comment by carlosparadis** [comment]
To be clear, Nico's task for Saturday is to finish Milestone 1 unit tests for the entire group. He should look through all the pull request comments (with hopefully your guidance so he knows what was already fixed and what is not), and should then make any remaining changes I asked so I can make a final review pass on them and hopefully integrate. 

For example, I believe Malia's unit test still says test_that("test that.....") etc. The unit tests should be normalized so they are consistent to Kaiaulu codebase. Hence, one person make revisions to all unit tests to ensure consistency so we don't end up in the first iteration of DV8 situation where Nico used rstudioapi, while you and Malia used system2 for example.

**Comment by leilani-reich** [comment]
Yes, those are the only unit tests for milestone 1. As for milestone 2, the bugzilla crawler unit test task is mentioned here https://github.com/sailuh/kaiaulu/issues/153#issue-1587260581 and the bugzilla crawler unit test task here https://github.com/sailuh/kaiaulu/issues/152.

Also, I believe Malia's most recent commit to this pull request addressed the "test that".

**Comment by carlosparadis** [comment]
Maybe I am getting confused about "Malia's revisions are done". I originally thought you were referring to the Bugzilla Notebook, but I guess you were just referring to Malia's Milestone 1 unit tests. Is that correct? 

What of yours and Nico's unit tests? Are all revisions done too? Meaning, this single pull request contains all unit tests of Milestone 1 with all revisions done and ready to merge? If this is the case, then Nico's next week's deadline becomes this weeks deadline: He should implement milestone 2 unit tests by Saturday, the 22nd. I can then decide on the subsequent week task at our meeting. The process and the deadlines remain the same, only the tasks will change. 

Let me know whether Nico still has revisions to do for Milestone 1, or otherwise, if he is finishing Milestone 2 unit tests by the 22nd. Either way, he should be assigned a deliverable for the 22nd if nothing is left of milestone 1.

**Comment by leilani-reich** [comment]
Yes, I'm just referring to the milestone 1 unit tests.
I think Nico's revisions for milestone 1 were mostly done as well, but I know my unit tests are not. I never got to making any revisions for my milestone 1 tests here https://github.com/sailuh/kaiaulu/pull/162.

Also, to be clear on associated milestone 1 pull requests:
- originally mine (I reassigned Nico) https://github.com/sailuh/kaiaulu/pull/162
- Malia's https://github.com/sailuh/kaiaulu/pull/163
- Nico's https://github.com/sailuh/kaiaulu/pull/167

**Comment by carlosparadis** [comment]
Thank you for making the references in a concise message. Then yes, this (your revisions to milestone 1 and any others of milestone 1) suffices for Nico's 22nd deliverable and/or ensuring your unit tests have similar formatting, as I exemplified between his code having rstudioapi (and also sprintf), and both yours and Malia have system2. He should ensure that across all Milestone 1 unit tests too.

What I will expect then from Nico is a commit to this Malia's pull request (via her fork) by the 22nd, finishing this, correct?

Thank you!

**Comment by nicoelee123** [comment]
All the revisions seemed to be made and the tests all pass using the 'Build' -> 'Test Package'. Please let me know if there's something that I missed, otherwise it looks good to me!

**Comment by carlosparadis** [comment]
@nicoelee123 Could you make a commit changing NEWS.md for me mentioning under Documentation that unit tests were added? You need to commit to Malia's fork, so she has to grant you permission to her fork to do that.

Once you do that, the GitHub Actions I set will also test a few more things for us here. Make sure you merge her code to the more recent versions of the master branch, otherwise it will not trigger the tests.

**Comment by nicoelee123** [comment]
@carlosparadis I think I updated correctly and made changes. It says there are some checks queued up as well.

**Comment by carlosparadis** [comment]
@nicoelee123 

If you click the "R-CMD-check" you can find the following in the text:
Screen Shot 2023-04-24 at 3 01 34 AM

As you can see, the tests are failing. Are they passing on your machine? This is why I added the GitHub actions. It is basically simulating what would happen if I were to download and run the unit test on mine.

**Comment by carlosparadis** [comment]
Hi Nico,

I added one more GitHub Action for test-coverage. This should give you some more additional info on why the checks failed. Here is one more example on reading them. This is from the test-coverage:

Screen Shot 2023-04-24 at 3 56 48 AM

It seems there is a problem in that region. 

Please feel free to do more commits until the checks pass, and if you can't understand why ping me. Also, 
make sure you do one more merge across all the PRs for unit tests so all of them have the more recent GitHub actions. This should hopefully be my last commit you need to merge to the PRs.

Thanks.

File: tests/testthat/test-git.R [hunk 13]
```diff
@@ -0,0 +1,22 @@
+test_that("Calling git_checkout with correct branch name and an exist local path of github project returns list of string", {
+  git_repo_path <- suppressWarnings(git_create_sample_log())
```
Comment (carlosparadis) [inline tests/testthat/test-git.R:2]: @nicoelee123 Could you confirm why suppressWarnings is needed here?

Also, this may be the function you need, instead of tryCatch, to handle the warning being generated on the problematic unit test in Leilani's PR.

**Comment by carlosparadis** [comment]
@nicoelee123 

Let me know when you make the fixes for:

https://github.com/sailuh/kaiaulu/pull/162#pullrequestreview-1407031000

And can check that the same issues don't happen here and I will take another pass. I scanned quickly, and at least the unit test descriptions of this PR looks fine.

Thanks!

File: tests/testthat/test-git.R (hunk 13 above)
Comment (nicoelee123) [reply tests/testthat/test-git.R:2]: @carlosparadis The suppressWarnings are there becuase the git_create_sample_log() and git_delete_sample_log() throws a warning everytime and I'm not sure why. I assumed it would be okay to suppress because despite the warnings, the sample log is successfully created and deleted. 

image

**Comment by nicoelee123** [comment]
@carlosparadis I made the changes and confirmed why the suppressWarnings were necessary.

**Comment by carlosparadis** [comment]
@nicoelee123 why were they necessary?

**Comment by nicoelee123** [comment]
The functions git_create_sample_log() and git_delete_sample_log() return a warning but it still successfully creates and deletes the sample log. I assumed based on Leilani's PR that you don't want warnings appearing, but if this warning is okay then I can remove the supressWarnings calls.

**Comment by carlosparadis** [comment]
No, it's fine. Keep them. I just wanted to make sure the reason is noted down! Thanks!

File: tests/testthat/test-parser.R [hunk 14]
```diff
@@ -5,6 +5,7 @@ test_that("Correct git repo path", {
 })
 test_that("Calling parse_gitlog with correct perceval and correct git log path returns a data table", {
   tools_path <- "../../tools.yml"
+  tools_path <- file.path(getwd(), tools_path)
```
Comment (carlosparadis) [inline tests/testthat/test-parser.R:8]: Hi Malia,

`getwd()` will not necessarily work. If the user changes their directory via setwd() this will fail too.  I am curious on why you added this too. Was `filepath(tools_path)` not working for you?

Make sure you

Comment (malialiu) [reply tests/testthat/test-parser.R:8]: It works, I'm just wondering if the errors happen because it didn't found the perceval path

Comment (carlosparadis) [reply tests/testthat/test-parser.R:8]: Just to be clear: Does it work without `getwd()`?

Comment (malialiu) [reply tests/testthat/test-parser.R:8]: Yes

Comment (carlosparadis) [reply tests/testthat/test-parser.R:8]: You are correct in that the test will fail via the Check for not finding Perceval. I am just trying to understand if we need the "getwd()" here.

Comment (malialiu) [reply tests/testthat/test-parser.R]: We don't need getwd()

**Comment by carlosparadis** [comment]
@malialiu there was one unit test that was failing for me, but my commit address it. However, after accepting the other unit tests I now have merge conflicts. Could fix these for me? Otherwise, this is ready to be accepted into master.

**Review by carlosparadis — APPROVED** [review]
Thanks, these pass on my local machine.
